### ISSUE TITLE
Implement online save storage

### ISSUE THREAD

**Issue by Ishadijcks** [pokeclicker/pokeclicker#340]
Implement online saving.

The user should be able to store their save in the cloud by entering credentials preferably only once.

When the user loads, they can enter their credentials again (if lost), and their savefile is restored again.

https://firebase.google.com/docs/auth
https://github.com/firebase/firebaseui-web

**Comment by greeny** [comment]
Based on a discussion in Discord server, I offer my help with this.

I don't know firebase, so I won't be able to help much with that, but my offer is to write (probably as a separate repository) a save storage server (most likely in PHP, as that's my "field of expertise"), which would provide a JSON REST API for the app to use.

Several suggestions if you'd go this way:
- user could get a randomly generated storage with assigned token (a random 32? chcaracter string), which would display in the UI for a user to copy-paste and save somewhere.
- game would periodically save new state to the cloud using this token. State would consist of the JSON data already used for local saving. There's no need to base64 encode that.
- in case user changes to a new PC/device, they could paste the token and link the device with the cloud storage, loading and overwriting locally saved data with the cloud data.
- the token would be saved in the local storage, allowing for data refresh on every load of the page (download from cloud).
- ~~we could also implement some "last change date/time", which would prevent overwriting the data from user if they e.g. played offline~~ probably better idea is "played time" in the save, as we would evade timezone problem with this. Is offline play even a thing? Or planned in the future?

This is like the simplest way I can think of to implement this, with possibility to extend it in multiple ways. Ofc we can do the whole "create account, forgot password, etc." workflow, but unless there's some advantage of this that I didn't see, it seems like an overkill.

Let me know if this is something you'd be interested in, if yes, I can even try and prepare some basic framework and code so you could start experimenting with a live prototype. Feel free to reach on Discord if you want futher discussion (greeny#4945).

**Comment by Ishadijcks** [comment]
> Is offline play even a thing? Or planned in the future?

Not at the moment.

But good ideas! I don't think I'll work on it *soon* but it is definitely nice to have.

**Comment by EmrysMyrddin** [comment]
@greeny Are you still working on it ?

@Ishadijcks Do you still want this feature ? What are your requirements ?

Do you have any deployment constraints ?

What is the expected number of users ?

**Comment by Qwertypop04** [comment]
> @greeny Are you still working on it ?
> 
> @Ishadijcks Do you still want this feature ? What are your requirements ?
> 
> Do you have any deployment constraints ?
> 
> What is the expected number of users ?

Neither of these people have been around for ages. Online saves has been ruled out, it's just not viable.

**Comment by Ishadijcks** [comment]
I've left this project a while ago, maybe @RedSparr0w can help :)

**Comment by EmrysMyrddin** [comment]
Do you mean it's not anymore in your roadmap or you're not even open for contributions on this topic ?

**Comment by Ishadijcks** [comment]
I am no longer working on PokeClicker since a few years ago, and transferred ownership to RedSparr0w.

You'll have to ask him!

**Comment by greeny** [comment]
> @greeny Are you still working on it ?
> 
> @Ishadijcks Do you still want this feature ? What are your requirements ?
> 
> Do you have any deployment constraints ?
> 
> What is the expected number of users ?

I'm still around contrary to what @Qwertypop04 thinks, however given that my suggestion generated pretty much no response when I posted it (there was some discussion in discord, but that's about it), I didn't really work on any prototype. My offer still stands though, if online saving would be added.

**Comment by Farboo171** [comment]
Per our discussion on Discord, there are no near term plans to do this, due to the difficulty of implementations, the potential attention from a host about our IP use it would attract, and the "Good enough for now" solution of DMing the discord bot. Closing.
